python/ray/data/_internal/datasource/kafka_datasource.py

```diff
@@ -0,0 +1,568 @@
+                partition_id: int = partition_id,
+                bootstrap_servers: List[str] = bootstrap_servers,
+                start_offset: Optional[Union[int, str]] = start_offset,
+                end_offset: Optional[Union[int, str]] = end_offset,
+                authentication: Dict[str, Any] = authentication,
+                timeout_ms: int = timeout_ms,
+            ):
+                """Create a Kafka read function with captured variables.
+
+                This factory function captures configuration variables as default arguments
+                to avoid serialization issues when the read function is executed remotely
+                by Ray. Using default arguments ensures all needed config is available
+                in the remote task without requiring 'self' to be serialized.
+                """
+
+                def kafka_read_fn() -> Iterable[Block]:
+                    """Read function for a single Kafka partition using kafka-python.
+
+                    This function runs remotely in a Ray task. It creates a KafkaConsumer,
+                    reads messages from a single assigned partition, and yields PyArrow tables
+                    incrementally for efficient streaming processing.
+                    """
+                    from kafka import KafkaConsumer, TopicPartition
+
+                    # Build consumer configuration
+                    consumer_config = _build_consumer_config_for_read(
+                        bootstrap_servers, authentication, topic_name, partition_id
+                    )
+
+                    # Create the Kafka consumer
+                    consumer = KafkaConsumer(**consumer_config)
+                    try:
+                        # Assign only the specific partition for this task
+                        topic_partition = TopicPartition(topic_name, partition_id)
+                        consumer.assign([topic_partition])
+
+                        # Determine start offset for this partition
+                        if start_offset is not None:
+                            start_off = _resolve_offset(
+                                consumer, topic_partition, start_offset, is_end=False
+                            )
+                        else:
+                            beginning_offsets = consumer.beginning_offsets(
+                                [topic_partition]
+                            )
+                            start_off = beginning_offsets[topic_partition]
+
+                        # Determine end offset for this partition
+                        if end_offset is not None:
+                            end_off = _resolve_offset(
+                                consumer, topic_partition, end_offset, is_end=True
+                            )
+                        else:
+                            end_offsets = consumer.end_offsets([topic_partition])
+                            end_off = end_offsets[topic_partition]
```

Comment:
Currently, the resolution logic is split between here and `_resolve_offset`. If you allow `_resolve_offset` to accept `None`, this code might be cleaner
```suggestion
                        # Determine start offset for this partition
                        start_off = _resolve_offset(
                            consumer, topic_partition, start_offset, is_end=False
                        )
                        
                        # Determine end offset for this partition
                        end_off = _resolve_offset(
                            consumer, topic_partition, end_offset, is_end=True
                        )
```

Reply:
I refactored `_resolve_offset`, should be much cleaner now.